tiledb/sm/query/query.h

```diff
@@ -193,6 +194,160 @@ class Query {
    */
   Status set_subarray(const void* subarray);
 
+  /**
+   * Return the query subarray
+   * @tparam T
+   * @return vector containing copy of the subarray
+   */
+  template <class T>
+  std::vector<T> subarray() const {
+    void* subarray = nullptr;
+    if (type_ == QueryType::WRITE) {
+      subarray = const_cast<Writer*>(&writer_)->subarray();
+    } else {  // READ
+      subarray = const_cast<Reader*>(&reader_)->subarray();
+    }
+
+    if (subarray == nullptr) {
+      LOG_STATUS(
+          Status::QueryError("Cannot get subarray; subarray is nullpyt"));
+      return {};
+    }
+
+    auto array_schema = this->array_schema();
+    if (array_schema == nullptr) {
+      LOG_STATUS(
+          Status::QueryError("Cannot get subarray; Array schema not set"));
+      return {};
+    }
+
+    // Get subarray size
+    uint64_t subarray_size = 2 * array_schema->coords_size();
+    uint64_t subarray_length =
+        subarray_size / datatype_size(array_schema->coords_type());
+
+    // Switch on domain type to validate the requested template matches the
+    // subarray datatype
+    switch (array_schema->domain()->type()) {
```

Comment:
Can you abstract the body of the switch cases into a templated helper function, similar to e.g. `Reader::dense_read()`? Otherwise the switch block should probably go in a separate function.